### DISCUSSION TITLE
File handling and working with Blobs

### DISCUSSION THREAD

**Discussion by Folcon** [day8/re-frame#729] [Q&A]
I've been using core.async and promises a lot to use the [FileReader](https://developer.mozilla.org/en-US/docs/Web/API/FileReader), [File](https://developer.mozilla.org/en-US/docs/Web/API/File) and [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob) api's.

However I do wonder if there's a better way of handling it that fit's more cleanly into how re-frame works.

One especially significant bugbear that I've not figured out how to get working is doing chunked reading.

Splitting up a longer task within re-frame and processing it in chunks is straightforward, but I've not figure out how to map this to the file api.

Any suggestions?

**Comment by superstructor** [comment]
A lot of browser APIs, such as async File related stuff, is promise-based so at some level you have to deal with the promises. I use [promesa](https://github.com/funcool/promesa).

In terms of integrating it into re-frame I think the best way is an effect handler (`reg-fx`) that contains the promise-based code and has on `on-chunk` and `on-success` or `on-complete` dispatches; e.g.

```clojure
(reg-fx
 :chunked-file-reader
(fn  [{:keys [on-chunk on-complete on-failure}]
  ...
  ;; on each chunk
  (dispatch (conj on-chunk the-chunk))
  ))

(reg-event-fx
  :read-file
  (fn [cofx event]
    {:chunked-file-reader {:on-chunk [:file-chunk]
                                        :on-complete [:file-complete]}}))
```

**Reply by olivergeorge** [reply]
I think this is conservative & viable but wonder if re-frame could provide a better alternative.

The common requirement here seems to be massaging inputs using async apis before invoking a handler.  Happily we don't need to lock app-db to do this work.

```
(rf/reg-event-fx2
   :file-upload 
   (fn signal [[_ file]] (.text file))
   (fn compute [ctx text] ...))
```

Huh, written like that it feels a lot like a reg-sub declaration.
